_P201_VBAN.ino

```diff
@@ -0,0 +1,577 @@
+#define VBAN_CODEC_UNDEFINED_11  0xD0 
+#define VBAN_CODEC_UNDEFINED_12  0xE0 
+#define VBAN_CODEC_USER   0xF0
+
+#define VBAN_PORT 6980 
+
+/* Attenuation
+When VDD_A is 3.3V:
+0dB attenuaton (ADC_ATTEN_0db) gives full-scale voltage 1.1V
+2.5dB attenuation (ADC_ATTEN_2_5db) gives full-scale voltage 1.5V
+6dB attenuation (ADC_ATTEN_6db) gives full-scale voltage 2.2V
+11dB attenuation (ADC_ATTEN_11db) gives full-scale voltage 3.9V (see note below)
+*/
+#define ADC_ATTENUATION ADC_ATTEN_6db
+//#define ADC_CHANNEL ADC1_CHANNEL_7
+adc1_channel_t ADC_CHANNEL = ADC1_CHANNEL_0;
+
+
+/*-----Internal variables and buffers ----- */
+uint8_t *buffer;  //a circular buffer for vban package with audio data  
+
+
+
+uint32_t frameCounter=0; //the incrementing frame counter for vban packages
+int32_t transmitNow = -1; // -1=no sending; otherwise we record here the starting point a multiple of VBAN_PACKET_SIZE
+
+//bool plugin_not_initialised=true;
+
+uint16_t audio_power=0;
+
+
+TaskHandle_t adcHandlerTask; //task to send UDP packages 
+hw_timer_t * timer = NULL; //timer used to aquire audio data
+portMUX_TYPE timerMux = portMUX_INITIALIZER_UNLOCKED; //mutex for the timer
+
+uint16_t sample_idx=VBAN_HEADER_SIZE; //index for audio aquisition; set to beginning of the first data portion
+
+WiFiUDP p201UDP; //the UDP object
+IPAddress UDP_IP;// = IPAddress(192,168,100,62);
+uint16_t UDP_PORT = VBAN_PORT;
+int16_t AUDIO_OFFSET=0;
+
+#define SAMPLES 256 //number of sampels per package
+#define CHANNELS 1 //number of channels
+//#define STREAM_NAME "Stream1" //name of the stream
+#define SAMPLE_SIZE_BYTES 2
+
+#define AUDIO_SIZE (SAMPLES * CHANNELS * SAMPLE_SIZE_BYTES)
+#define VBAN_PACKET_SIZE  (VBAN_HEADER_SIZE + AUDIO_SIZE)
+#define BLOCKS 2 
+#define P201_BUFFER_SIZE (BLOCKS * VBAN_PACKET_SIZE)
+
+void P201_fillHeader(struct tagVBAN_HEADER * header){
+  strcpy_P(header->vban, PSTR("VBAN")); 
+  header->format_SR = VBAN_PROTOCOL_AUDIO | 8; // 9; //32KHz or use  8; //16Khz or use 7; //8Khz
+  header->format_nbs = SAMPLES-1;  
+  header->format_nbc = CHANNELS-1; //mono   
+  header->format_bit = VBAN_CODEC_PCM | VBAN_DATATYPE_INT16;    /* mask = 0x07 (nb Byte integer from 1 to 4) */         
+  strcpy_P(header->streamName , Settings.Name );/* stream name max 16 chars*/         
+  header->nuFrame = 0;       /* growing frame number. */     
+}
+
+void P201_setFrameCounter(uint8_t *header, uint32_t counter){
+ *(header+24) = (uint8_t)counter; //LSB
+ *(header+25) = (uint8_t)(counter>>8);
+ *(header+26) = (uint8_t)(counter>>16);
+ *(header+27) = (uint8_t)(counter>>24); //MSB
+}
+
+#define ismulticast(addr1) (((uint32_t)(addr1) & PP_HTONL(0xf0000000UL)) == PP_HTONL(0xe0000000UL))
+
+
+int P201_sendUDP(IPAddress toIP, uint16_t port, uint8_t * package, size_t length){
+  if (WiFiConnected()) 
+  {
+    if (port < 0 || port > 65535) return false;
+    if (ismulticast(toIP))
+    {
+      if (p201UDP.beginMulticastPacket()) //(toIP, port, WiFi.localIP(),2))

```

Comment:
Just curious. How well does multicast work on WiFi?

Reply:
does not work yet. still working on it.

Reply:
Well, the reason I asked is because I was surprised it would even work via WiFi.
So I would not put too much effort in it to make it work at all.

Reply:
It should work as only one packet will be sent like in unicast.
There can be many receivers that will subscribe to this special address.

Reply:
Yep, that's the theory of multicast.
But I've not yet seen an implementation that would work well when the subscribed units also connected to the same WiFi access point.
So that's why I was curious whether it would work well.